### ISSUE TITLE
Dapr lifecycle management improve - start

### ISSUE THREAD

**Issue by skyao** [dapr/dapr#2918]
## In what area(s)?

/area runtime

## Background

We, Allbaba Cloud, are trying to use dapr in our production. We found many errors like "dapr unavailable" in our logs so that we want to improve the dapr lifeciecle management between App and daprd.

Because there are so many detais, I split the proposal to three parts: start / stop and running.  

In this issue we will focus on the first part : start.

## user cases when app and daprd start

It's very complicated for start scenario because there are different user case for outbound requests and inbound requests. I will talk about them one by one.

### For outbound Requests

**case1: App sends outbound requests after starts**

Description: When App is started, it perhaps sends some outbound requests imediately. So at this time, the daprd should be able to serve these outbound requests from App. 

Requirement: Daprd should be ready for outbound before App is ready

**case2: App loads configurations during starting**

Description: We are planning to add configuration building block in Dapr. When App is starting, it will load configuration through daprd . So the daprd should be able to serve these outbound requests (loading configurations) from App. 

Requirement: Daprd should be ready for outbound when App is starting.

### For inbound Requests

**case3: Daprd forwards inbound requests to App**

Description: When Daprd is started and subscribed some topic, it will receive inbound requests imediately and forward them to App. So at this time, the App should be able to serve these inbound requests. 

Requirement: App should be ready for inbound before Daprd starts to subscribe and forward inbound requests.

**case4: Daprd gets configuration from App during starting**

Description: During starting, Daprd need to get some dapr related configuration from App like topics to subscribe/input binding to read. So App should be ready to serve dapr's request before Daprd start to do subscribe .

Requirement: App should be ready before Daprd start to get configuration for  inbound.

### Proposal Overview

From above user cases we know that daprd and app need to wait for each other:

- for outbound: Daprd should be ready before app
- for inbound: App should be ready before daprd

So we can't simply let Daprd start first or let App start first. We need to split daprd start process into two part and there are two milestones:  1. Daprd is ready for serve outbound traffic (short for outbound ready) 2. Daprd is ready to subscribe and forward inbound requests (short for inbound ready) . And the process from App start to App ready should between these two part of Daprd.

As shown in following timeline picture, there are three phases for the start of Daprd and App:

![image](https://user-images.githubusercontent.com/1582369/110908217-74c74600-8349-11eb-8832-0a3cb2b41ac5.png)

- Phase1: daprd starts and becomes to outbound ready. App should not start during this phase.  We need to block the App start process until daprd is outbound ready.
- Phase2: App starts and becomes to ready. Daprd should be blocked during this phase until App is ready.
- Phase3: Daprd continue to start and become to inbound ready.

Now with this design of three phases start process and two blocks between the three phases, we can fulfill the requirements for user case 1/2/3/4.

To implement it, we need more details, as shown in this picture:

![image](https://user-images.githubusercontent.com/1582369/110908284-8b6d9d00-8349-11eb-9cef-e28d9f3af9f7.png)

The key points are to implement the two blocks.

### Block1: block App until daprd is outbound ready

In kubenetes, to block app container and let daprd start first, we need to do:

1. make daprd as first container in pod: Yaron had done this in issue [Add dapr sidecar as first container in pod #2341](https://github.com/dapr/dapr/pull/2341)

2. use a postStart hook to block other containers

	we can reference istio's practices as described in this issue [Allow users to delay application start until proxy is ready #24737](https://github.com/istio/istio/pull/24737)
	
	> for startup issues, the istio community came up with a quite clever workaround which basically injects envoy as the first container in the container list and adds a postStart hook that checks and wait for envoy to be ready. This is blocking and the other containers are not started making sure envoy is there and ready before starting the app container.

For daprd, the implementation will be a little complicated: in istio, app waits for envoy ready; in dapr, App waits for daprd's outbound ready, not totally ready. So we can't use the healthz endpoint of daprd directly, we need to add new special end point for "dapr outbound ready" statue:

- Add a new statue isOutboundReady in daprd 
- Add a new end point like "http://localhost:3500/v1.0/healthz/outbound", when isOutboundReady is true, response with 200.
- In daprd, mark isOutboundReady to true in runtime initial: after the components loaded and api server started, before blockUntilAppIsReady() method
- in the postStart hook, check the end point of OutboundReady, block until it responses 200

### Block2: block Daprd until App is ready

We have a blockUntilAppIsReady() method in daprd for this, it will block daprd during dapr runtime initial. This is exactly what we need.

But the code implementation of blockUntilAppIsReady() method is just to try to dial a tcp connection:

```
func (a *DaprRuntime) blockUntilAppIsReady() {
   for {
      conn, _ := net.DialTimeout("tcp", net.JoinHostPort("localhost", fmt.Sprintf("%v", a.runtimeConfig.ApplicationPort)), time.Millisecond*500)
      if conn != nil {
         conn.Close()
         break
      }
      time.Sleep(time.Millisecond * 50)
   }
}
```

There is a risk between "app port is open" and "app is ready".

To be safe, we suggest add a logic in blockUntilAppIsReady() method to check by the readiness of app container:

- Add an env variable "APP_READINESS_ADDRESS" for daprd container
- if this variable is set, try to do a readiness checking by sending a HTTP Get to this address. 
- This readiness checking is optional: if "APP_READINESS_ADDRESS" is not set, we will fallback to check TCP connection.

**Comment by xiazuojie** [comment]
@skyao The problem you described is valid. However, the proposal you give feels way too complicated to implement, and it's not solving the root cause of the problem: why does Daprd and app seem depend on each other? Why are their life cycles mixed up? Does it have to be this way?

## why does Daprd and app seem depend on each other?
The root cause is both the app and daprd are designed to be **allowed** to be stateful during start-up. 
* The app: 
1. App sends outbound requests after starts
2. App loads configurations during starting
* daprd: 
1.  Daprd forwards inbound requests to App
2. Daprd gets configuration from App during starting

## Does it have to be this way?
Absolutely not. Daprd can be designed to be required to be stateless during startup.
* change "Daprd forwards inbound requests to App" to "app asks Daprd to start binding some inbound requests after the app has started and is ready". 
* change "Daprd gets configuration from App during starting" to "App pushes its configuration to daprd after the app has started and is ready"

Basically, daprd itself does not have any configuration and thus is stateless during startup. All configuration comes from the app during runtime.

With this design, daprd can always start first, and then the app. It's a one way dependency.

**Comment by skyao** [comment]
@xiazuojie the problem of darpd start is that daprd need to support both outbound traffic and inbound traffic. So anyway we need this three-phases-start to prepare for outbound and inbound.

Your solution still need three-phases-start:

- Phase1: daprd starts and becomes to ready. (in your solution daprd will directly become to ready, but in fact only outbound ready)
- Phase2: App starts and becomes to ready. 
- Phase3: ~~Daprd continue to start and become to inbound ready.~~. App triggers daprd to continue to start for inbound ready.

The difference is that :

- in phase1 we can use daprd healthz enpoint directly in postStart hook
- in phase3 App will take the responsibility of trigger daprd to start inbound traffic with the configutations. The second blocking now can removed, but the developer of app need to do this trigger.

I agree to your suggestion to simplize the dependency between daprd and app, but it will change current dapr design and broken current dapr api.

**Comment by xiazuojie** [comment]
> @xiazuojie the problem of darpd start is that daprd need to support both outbound traffic and inbound traffic. So anyway we need this three-phases-start to prepare for outbound and inbound.
> 
> Your solution still need three-phases-start:
> 
> * Phase1: daprd starts and becomes to ready. (in your solution daprd will directly become to ready, but in fact only outbound ready)
> * Phase2: App starts and becomes to ready.
> * Phase3: ~Daprd continue to start and become to inbound ready.~. App triggers daprd to continue to start for inbound ready.
> 
> The difference is that :
> 
> * in phase1 we can use daprd healthz enpoint directly in postStart hook
> * in phase3 App will take the responsibility of trigger daprd to start inbound traffic with the configutations. The second blocking now can removed, but the developer of app need to do this trigger.
> 
> I agree to your suggestion to simplize the dependency between daprd and app, but it will change current dapr design and broken current dapr api.

"App triggers daprd to continue to start for inbound ready". Yes it's still needed, but it does not have to be part of the "starting" process. The app can ask for input binding at any time - at the start or during runtime.

**Comment by nobodyiam** [comment]
I agree with @xiazuojie that from the life cycle point of view, daprd should not depend on the app during startup.

**Comment by yaron2** [comment]
> I agree with @xiazuojie that from the life cycle point of view, daprd should not depend on the app during startup.

Dapr needs to wait until the app is ready in the following cases:

1. The app subscribed to incoming events (with either pub/sub or input bindings): In this case, Dapr should not start consuming events before the app is ready to receive them.
2. The app is being called using Service Invocation.

Today, Dapr honors the two cases above by starting to consume events and setting up Service Invocation _after_ the app is listening on a port.

@skyao I agree that adding an *optional* health probe on the app is a good way to do that.

Regarding the sidecar ready to receive requests before the app:

Today, the secrets API and state API are both ready to be consumed before the sidecar is marked as ready. This is done to allow apps to get things like configuration from a state store before starting to listen for incoming requests (and thus marking the sidecar as "total ready").

However, there isn't a good way for the app today to check which APIs are ready and which are not.

**Comment by nobodyiam** [comment]
@yaron2 

> Dapr needs to wait until the app is ready in the following cases:
> 
> 1. The app subscribed to incoming events (with either pub/sub or input bindings): In this case, Dapr should not start consuming events before the app is ready to receive them.
> 2. The app is being called using Service Invocation.
> 
> Today, Dapr honors the two cases above by starting to consume events and setting up Service Invocation _after_ the app is listening on a port.

This is a necessary step. But IMO it doesn't belong to the _start_ lifecycle of dapr since dapr also need to handle the situation when the app crashes after startup. Imagine we draw a state diagram, then this would be the state transition between states like ready and not ready, but it won't change the state to starting.

@skyao Speaking of this, I think it would be more clear to describe the lifecycle using a [state diagram](https://en.wikipedia.org/wiki/State_diagram).

**Comment by skyao** [comment]
> I agree with @xiazuojie that from the life cycle point of view, daprd should not depend on the app during startup.

I think this is a big change for dapr, should we consider it in this lifecycle issue? Or should we open another issue to discuss? I would like to go on this proposal based on current dapr design, and if later dapr changes to not depending on app, I can update it.

I don't want to block this lifecycle issue any more, since there are still many error logs in our production env.

@yaron2 Can I raise a PR for this fix now?

**Comment by yaron2** [comment]
>  since dapr also need to handle the situation when the app crashes after startup

Lets break down Dapr's behavior today when the app crashes and the sidecar is running:

1. Service Invocation: calling apps will get unavailable errors. Dapr sidecars on the caller side will retry, and if retry attempts exceed threshold, the caller app will get an error and can then retry on its own. This is pretty much aligned with any request-response scenario as far as the caller is concerned.
2. Pub/Sub - Dapr will not be able to deliver messages to the app, and those will get NACKed by Dapr and sent to another healthy instance or retried later
3. Bindings - Same as pub/sub

Considering the above, I feel like we might not need a special check and different tactics when the app itself is down.

@skyao I agree, you can go ahead and submit a PR. @nobodyiam we will continue discussing this here, and perhaps create chaos tests with fault injection to be able to observe the different situations discussed here.

**Comment by nobodyiam** [comment]
@yaron2 

Let me try to explain my point via some diagrams.
Suppose we have 2 applications App1 and App2:
* App1 calls App2
* App1 has one instance with ip 1.1.1.1
* App2 has two instances with ip 2.2.2.2 and 3.3.3.3

## Normal situation
Under normal situation, 2.2.2.2 and 3.3.3.3 will receive 50% of the total traffic, respectively.

![image](https://user-images.githubusercontent.com/837658/112411353-ba2b4080-8d57-11eb-8353-0d14fc96da3d.png)

## Abnormal situation
Now suppose 2.2.2.2 is crashed, then how much traffic will 2.2.2.2 and 3.3.3.3 receive?

### Option 1
2.2.2.2 and 3.3.3.3 still receives 50% of the total traffic, respectively.

![image](https://user-images.githubusercontent.com/837658/112412733-f65fa080-8d59-11eb-98c1-702d2f89cc46.png)

### Option 2
3.3.3.3 receives 100% of the total traffic.

![image](https://user-images.githubusercontent.com/837658/112412752-feb7db80-8d59-11eb-9364-32fd2f0047e0.png)

I haven't tested this scenario since dapr cli doesn't allow me to start 2 dapr instances with the same appid on my laptop, but from my point of view, option 2 is more reasonable.
If option 2 is the current implementation, then it means dapr instance would check the application's health status not only during start phase but also when it's running.

**Comment by yaron2** [comment]
@skyao are the linked PRs enough to close this issue?
 
If not, whats left?

**Comment by skyao** [comment]
Block1: block App until daprd is outbound ready
State： done after PR #3052  and #3150 

Block2: block Daprd until App is ready
State：need to improve.

Is this solution I mentioned in the proposal OK? 

> There is a risk between "app port is open" and "app is ready".
>
> To be safe, we suggest add a logic in blockUntilAppIsReady() method to check by the readiness of app container:
>
> - Add an env variable "APP_READINESS_ADDRESS" for daprd container
> - if this variable is set, try to do a readiness checking by sending a HTTP Get to this address.
> - This readiness checking is optional: if "APP_READINESS_ADDRESS" is not set, we will fallback to check TCP connection.

**Comment by daixiang0** [comment]
@skyao what is the solution for "block Daprd until App is ready"?

**Comment by cshi0** [comment]
@daixiang0 Hi！We've proposed a process of checking the availability of the app in these issues: #3376 and #3377

**Comment by yaron2** [comment]
Reopened

**Comment by theonefx** [comment]
@yaron2

**Comment by trondhindenes** [comment]
We're working on some long-running requests at tme moment, so I'll leave some notes here:
- during startup, dapr consideres the app healthy once it responds on its designated port. This might not always be desirable. Dapr should offer the ability to use the app container's readiness state health check (if confidered) instead of simple port pinging. This would tie nicely into native kubernetes constructs, and make dapr a "nice citizen".
- we're seeing cases where dapr itself will restart during startup if it sidecars a slow-starting app. dapr's health check (which is auto-configured by the sidecar injector) should be configured with the same initial delay time as the app container, so that kubernetes doesn't start health-checking the dapr sidecar before its ready

- during shutdown for apps, the app container might take a minute or more to shut down if it's in the process of processing a message. If the app needs to place a message on the queue after (or as part of) processing a pubsub message, it needs the dapr sidecar available. However, during shutdown, the dapr sidecar responds immediately to the SIGTERM signal from kubernetes. This causes message loss for us. The correct behavior would be to have dapr wait to shutdown until the app container is killed and/or the app no longer listens on its designated port. During the shutdown phase dapr should not pick new messages from pubsub components (or respond to service invocations), it should only have its "outbound paths" enabled
